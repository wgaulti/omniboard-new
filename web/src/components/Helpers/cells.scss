.editable-cell {
  margin-top: -15px;
  margin-left: -8px;
}

.select-cell {
  margin-top: -15px;
  margin-left: -5px;
}

.delete-icon {
  &:hover {
    color: red;
  }
}

.selection-cell-wrapper {
  .checkbox label {
    padding-left: 0;
    line-height: 0;
  }

  .checkbox label:after {
    content: '';
    display: table;
    clear: both;
  }

  .checkbox .cr {
    position: absolute;
    top: 50%;
    left: 50%;
    display: inline-block;
    border: 1px solid #a9a9a9;
    border-radius: .30em;
    width: 1.5em;
    height: 1.5em;
    float: left;
    margin: -0.75em 0 0 -0.75em; /* Half of height 0 0 Half of width */
  }

  .checkbox .cr .cr-icon, .checkbox .cr .cr-indeterminate-icon {
    position: absolute;
    font-size: 0.9em;
    line-height: 0;
    top: 50%;
    left: 15%;
  }

  .checkbox label input[type="checkbox"] {
    display: none;
  }

  .checkbox label input[type="checkbox"] + .cr {
    & > .cr-icon, & > .cr-indeterminate-icon {
      opacity: 0;
    }
  }

  .checkbox label input[type="checkbox"]:checked + .cr {
    background-color: #337ab7;
    color: #FFFFFF;
    border: 1px #FFFFFF;

    & > .cr-icon {
      opacity: 1;
    }
  }

  .checkbox label input[type="checkbox"]:indeterminate + .cr {
    background-color: #337ab7;
    color: #FFFFFF;
    border: 1px #FFFFFF;

    & > .cr-icon {
      opacity: 0;
    }

    & > .cr-indeterminate-icon {
      opacity: 1;
    }
  }

  .checkbox label input[type="checkbox"]:disabled + .cr {
    opacity: .5;
  }
}
