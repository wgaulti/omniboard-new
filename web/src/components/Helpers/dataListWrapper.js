class DataListWrapper {
  constructor(data = [], size, end, fetchCallback) {
    this._data = data;
    // When fetching we need to fetch the index missing + additional x-elements.
    // This specifies that we add 10% of the total size when fetching, the
    // maximum number of data-requests will then be 10.
    this._fetchSize = DataListWrapper._getFetchSize(size);
    this._end = end;
    this._pending = false;
    this._size = size;
    this._fetchCallback = fetchCallback;
  }

  static _getFetchSize(size) {
    return isNaN(size) ? 0 : Math.ceil(Number(size) / 10);
  }

  getSize() {
    return this._size || 0;
  }

  fetchRange(end) {
    if (this._pending) {
      return;
    }

    this._pending = true;
    this._fetchCallback(end)
      .then(({data, count}) => {
        this._pending = false;
        this._end = end;
        this._size = count;
        this._data = data;
        this._fetchSize = DataListWrapper._getFetchSize(count);
      })
      .catch(error => {
        this._pending = false;
        console.error(error);
      });
  }

  getObjectAt(index) {
    // Fetch data 15 rows ahead of end.
    // This will improve user experience such that the user
    // will not see pending status for a long time.
    const numRowsAhead = 15;
    if (index + numRowsAhead >= this._end && this._end < this.getSize()) {
      this.fetchRange(Math.min(this.getSize(),
        index + numRowsAhead + this._fetchSize));
      return index < this._end ? this._data[index] : null;
    }

    return this._data[index];
  }

  getDataArray() {
    return this._data;
  }

  setObjectAt(index, value) {
    this._data[index] = value;
  }

  set data(value) {
    this._data = value;
  }
}

export {DataListWrapper};
