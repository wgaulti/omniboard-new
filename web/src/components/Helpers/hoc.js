import PropTypes from 'prop-types';
import React from 'react';
import './hoc.scss';

export class ProgressWrapper extends React.PureComponent {
  static propTypes = {
    children: PropTypes.node,
    loading: PropTypes.bool
  };

  render() {
    const {loading, children} = this.props;
    let html = children;
    if (loading) {
      html = (
        <div className='table-display'>
          <div className='table-cell-display gray'>
            <i className='glyphicon glyphicon-refresh glyphicon-refresh-animate'/> Loading...
          </div>
        </div>
      );
    }

    return html;
  }
}
