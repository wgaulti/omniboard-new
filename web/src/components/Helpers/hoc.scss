.table-display {
  display: table;
  width: 100%;
  height: 100%;
  .table-cell-display {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
  }
}

.gray {
  color: #808080;
}
