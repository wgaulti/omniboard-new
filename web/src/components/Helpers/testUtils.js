import mockAxios from 'jest-mock-axios';

export const generateMockResponse = (status, times) => {
  for (let i = 0; i < times; i++) {
    mockAxios.mockResponse({status});
  }
};
