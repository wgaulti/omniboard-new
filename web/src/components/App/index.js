import React, {Component} from 'reactn';
import {Navbar, Nav, MenuItem, NavDropdown, Glyphicon, NavItem} from 'react-bootstrap';
import axios from 'axios';
import PropTypes from 'prop-types';
import {ToastContainer, toast} from 'react-toastify';
import moment from 'moment-timezone';
import backend, {setDbInfo} from '../Backend/backend';
import {parseServerError} from '../Helpers/utils';
import 'react-toastify/dist/ReactToastify.min.css';
import './style.scss';
import RunsTable from '../RunsTable/runsTable';
import {
  SETTING_TIMEZONE,
  AUTO_REFRESH_INTERVAL,
  DEFAULT_AUTO_REFRESH_INTERVAL,
  INITIAL_FETCH_SIZE, DEFAULT_INITIAL_FETCH_SIZE, ROW_HEIGHT, DEFAULT_ROW_HEIGHT
} from '../../appConstants/app.constants';

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showCustomColumnModal: false,
      otherDbs: [],
      dbInfo: {},
      showSettingsModal: false,
      appVersion: ''
    };
  }

  static propTypes = {
    match: PropTypes.object,
    location: PropTypes.shape({
      search: PropTypes.string
    }),
    history: PropTypes.shape({
      push: PropTypes.func
    })
  };

  _resetCache = () => {
    localStorage.clear();
    location.reload();
  };

  _showCustomColumnModal = () => {
    this.setState({
      showCustomColumnModal: true
    });
  };

  _handleCustomColumnModalClose = () => {
    this.setState({
      showCustomColumnModal: false
    });
  };

  _showSettingsModal = () => {
    this.setState({
      showSettingsModal: true
    });
  };

  _handleSettingsModalClose = () => {
    this.setState({
      showSettingsModal: false
    });
  };

  /**
   * SettingsResponse is of type
   * [
   *  {
   *   "_id": "5ca1ce94a686ac25c4a78eea",
   *   "name": "timezone",
   *   "value": "America/Los_Angeles"
   *  }
   * ]
   * @param {array} settingsResponse response for global settings from API.
   * @private
   */
  _updateGlobalSettings = settingsResponse => {
    const settings = settingsResponse.reduce((acc, current) => {
      return {...acc, [current.name]: current};
    }, this.global.settings);

    this.setGlobal({
      settings
    });
  };

  _initializeSetting = (setting, value) => {
    backend.post('api/v1/Omniboard.Settings', {
      name: setting,
      value
    }).then(response => {
      if (response.status === 201) {
        this._updateGlobalSettings([response.data]);
      }
    });
  };

  _fetchData = () => {
    axios.all([
      axios.get('/api/v1/databases'),
      backend.get('api/v1/database'),
      backend.get('api/v1/Omniboard.Settings'),
      backend.get('api/v1/Version')
    ]).then(axios.spread((allDbResponse, dbResponse, settingsResponse, versionResponse) => {
      if (allDbResponse && allDbResponse.data) {
        this.setState({
          otherDbs: allDbResponse.data
        });
      }

      if (dbResponse && dbResponse.data) {
        this.setState({
          dbInfo: dbResponse.data
        });
      }

      if (versionResponse && versionResponse.data && versionResponse.data.version) {
        this.setState({
          appVersion: `v${versionResponse.data.version}`
        });
      }

      // Write default settings to the database for the first time
      // Guess the client timezone and set it as default
      const userTimezone = moment.tz.guess();
      if (settingsResponse && settingsResponse.data && settingsResponse.data.length > 0) {
        const settingsResponseData = settingsResponse.data;
        this._updateGlobalSettings(settingsResponseData);
        if (!settingsResponseData.some(setting => setting.name === SETTING_TIMEZONE)) {
          this._initializeSetting(SETTING_TIMEZONE, userTimezone);
        }

        if (!settingsResponseData.some(setting => setting.name === AUTO_REFRESH_INTERVAL)) {
          this._initializeSetting(AUTO_REFRESH_INTERVAL, DEFAULT_AUTO_REFRESH_INTERVAL);
        }

        if (!settingsResponseData.some(setting => setting.name === INITIAL_FETCH_SIZE)) {
          this._initializeSetting(INITIAL_FETCH_SIZE, DEFAULT_INITIAL_FETCH_SIZE);
        }

        if (!settingsResponseData.some(setting => setting.name === ROW_HEIGHT)) {
          this._initializeSetting(ROW_HEIGHT, DEFAULT_ROW_HEIGHT);
        }
      } else if (settingsResponse && settingsResponse.status === 200) {
        // If empty response, then initialize all settings
        this._initializeSetting(SETTING_TIMEZONE, userTimezone);
        this._initializeSetting(AUTO_REFRESH_INTERVAL, DEFAULT_AUTO_REFRESH_INTERVAL);
        this._initializeSetting(INITIAL_FETCH_SIZE, DEFAULT_INITIAL_FETCH_SIZE);
        this._initializeSetting(ROW_HEIGHT, DEFAULT_ROW_HEIGHT);
      }
    })).catch(error => {
      toast.error(parseServerError(error));
    });
  };

  componentDidMount() {
    const {match: {params}} = this.props;
    setDbInfo(backend, {path: params.dbPath});
    this._fetchData();
  }

  render() {
    const {showCustomColumnModal, showSettingsModal, dbInfo, otherDbs, appVersion} = this.state;
    if (!dbInfo || !dbInfo.key) {
      return <span>Loading app...</span>;
    }

    const localStorageKey = `${dbInfo.key}|RunsTable|1`;
    return (
      <div className='App'>
        <Navbar inverse fluid>
          <Navbar.Header>
            <Navbar.Brand>
              <a href='#'>Omniboard <span className='version'>{appVersion}</span></a>
            </Navbar.Brand>
          </Navbar.Header>
          <Navbar.Collapse>
            <Nav pullLeft activeKey={dbInfo.key}>
              {otherDbs.map(db => {
                return <NavItem key={db.key} eventKey={db.key} href={db.path}>{db.key} ({db.name})</NavItem>;
              })}
            </Nav>
            <Nav pullRight>
              <NavDropdown eventKey={1} title={<Glyphicon glyph='cog'/>} id='settings'>
                <MenuItem test-attr='reset-cache-button' eventKey={1.1} onClick={this._resetCache}>
                  <Glyphicon glyph='refresh'/>
                  &nbsp; Reset Cache
                </MenuItem>
                <MenuItem test-attr='manage-config-columns-button' eventKey={1.2} onClick={this._showCustomColumnModal}>
                  +/- Custom Columns
                </MenuItem>
                <MenuItem test-attr='settings-button' eventKey={1.3} onClick={this._showSettingsModal}>
                  <Glyphicon glyph='wrench'/>
                  &nbsp; Settings
                </MenuItem>
              </NavDropdown>
            </Nav>
          </Navbar.Collapse>
        </Navbar>
        <div className='content'>
          <ToastContainer autoClose={false}/>
          <RunsTable dbInfo={dbInfo} localStorageKey={localStorageKey}
            showCustomColumnModal={showCustomColumnModal}
            handleCustomColumnModalClose={this._handleCustomColumnModalClose}
            showSettingsModal={showSettingsModal}
            handleSettingsModalClose={this._handleSettingsModalClose}
            location={this.props.location} history={this.props.history}/>
        </div>
      </div>
    );
  }
}

export default App;
