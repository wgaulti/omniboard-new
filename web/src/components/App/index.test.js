import React from 'reactn';
import mockAxios from 'jest-mock-axios';
import {toast} from 'react-toastify';
import {parseServerError} from '../Helpers/utils';
import * as appConstants from '../../appConstants/app.constants';
import App from '.';

describe('App component', () => {
  let wrapper = null;
  toast.error = jest.fn();

  beforeEach(async () => {
    const app = <App match={{params: {dbPath: 'test'}}}/>;
    wrapper = shallow(app);
    wrapper.setState({dbInfo: {key: 'default', name: 'test_db'}});
    await tick();
  });

  afterEach(() => {
    mockAxios.reset();
    jest.clearAllMocks();
    // Reset localStorage
    localStorage.clear();
    React.resetGlobal();
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should reset cache', () => {
    // Mock location.reload method
    // eslint-disable-next-line no-unused-vars
    const {location} = window;
    delete window.location;
    window.location = {reload: jest.fn()};
    const value = 'testValue';
    localStorage.setItem('test', value);

    expect(localStorage.__STORE__.test).toEqual(value);
    wrapper.find('[test-attr="reset-cache-button"]').simulate('click');

    expect(Object.keys(localStorage.__STORE__)).toHaveLength(0);
    expect(window.location.reload).toHaveBeenCalledTimes(1);
  });

  it('should show/hide ManageCustomColumns modal', () => {
    wrapper.find('[test-attr="manage-config-columns-button"]').simulate('click');

    expect(wrapper.state().showCustomColumnModal).toBeTruthy();
    wrapper.instance()._handleCustomColumnModalClose();

    expect(wrapper.state().showCustomColumnModal).toBeFalsy();
  });

  it('should show/hide Settings modal', () => {
    wrapper.find('[test-attr="settings-button"]').simulate('click');

    expect(wrapper.state().showSettingsModal).toBeTruthy();
    wrapper.instance()._handleSettingsModalClose();

    expect(wrapper.state().showSettingsModal).toBeFalsy();
  });

  describe('should fetch database names on mount', () => {
    it('and handle success', async () => {
      expect(mockAxios.get).toHaveBeenCalledWith('api/v1/database');
      mockAxios.mockResponse({status: 200, data: [{key: 'default', path: '/test', name: 'test_db'}]});
      mockAxios.mockResponse({status: 200, data: {key: 'default', name: 'test_db'}});
      mockAxios.mockResponse({status: 200, data: [{name: 'timezone', value: 'Atlantic/Reykjavik', _id: 1}]});
      mockAxios.mockResponse({status: 200, data: {version: '1.0.0'}});
      await tick();

      expect(wrapper.update().state().otherDbs).toEqual([{key: 'default', path: '/test', name: 'test_db'}]);
      expect(wrapper.update().state().dbInfo).toEqual({key: 'default', name: 'test_db'});
      expect(wrapper.update().state().appVersion).toEqual('v1.0.0');
    });

    it('and handle error', async () => {
      const error = {status: 500, message: 'Unknown error'};
      mockAxios.mockError(error);
      mockAxios.mockError(error);
      await tick();

      expect(toast.error).toHaveBeenCalledWith(parseServerError(error));
    });
  });

  describe('should write default settings for the first time', () => {
    it('when none of the settings are present', async () => {
      const setting = {name: 'timezone', value: 'Atlantic/Reykjavik', _id: 1};
      const autoRefreshSetting = {name: 'auto_refresh_interval', value: 30, _id: 2};
      const initialFetchSize = {name: 'initial_fetch_size', value: 50, _id: 3};
      const rowHeight = {name: 'row_height', value: 70, _id: 4};
      mockAxios.mockResponse({status: 200, data: [{key: 'default', path: '/test', name: 'test_db'}]});
      mockAxios.mockResponse({status: 200, data: {key: 'default', name: 'test_db'}});
      mockAxios.mockResponse({status: 200, data: []});
      mockAxios.mockResponse({status: 200, data: {version: '1.0.0'}});
      await tick();

      expect(mockAxios.post).toHaveBeenCalledTimes(4);
      mockAxios.mockResponse({status: 201, data: setting});
      mockAxios.mockResponse({status: 201, data: autoRefreshSetting});
      mockAxios.mockResponse({status: 201, data: initialFetchSize});
      mockAxios.mockResponse({status: 201, data: rowHeight});

      await tick();

      expect(wrapper.update().instance().global.settings[appConstants.SETTING_TIMEZONE]).toEqual(setting);
      expect(wrapper.update().instance().global.settings[appConstants.AUTO_REFRESH_INTERVAL]).toEqual(autoRefreshSetting);
      expect(wrapper.update().instance().global.settings[appConstants.INITIAL_FETCH_SIZE]).toEqual(initialFetchSize);
      expect(wrapper.update().instance().global.settings[appConstants.ROW_HEIGHT]).toEqual(rowHeight);
    });

    it('when only auto refresh setting is present', async () => {
      const setting = {name: 'timezone', value: 'Atlantic/Reykjavik', _id: 2};
      const initialFetchSize = {name: 'initial_fetch_size', value: 50, _id: 3};
      const rowHeight = {name: 'row_height', value: 70, _id: 4};
      const autoRefreshSetting = {
        name: appConstants.AUTO_REFRESH_INTERVAL,
        value: 60,
        _id: 1
      };
      mockAxios.mockResponse({status: 200, data: [{key: 'default', path: '/test', name: 'test_db'}]});
      mockAxios.mockResponse({status: 200, data: {key: 'default', name: 'test_db'}});
      mockAxios.mockResponse({status: 200, data: [autoRefreshSetting]});
      mockAxios.mockResponse({status: 200, data: {version: '1.0.0'}});
      await tick();

      expect(mockAxios.post).toHaveBeenCalledTimes(3);
      mockAxios.mockResponse({status: 201, data: setting});
      mockAxios.mockResponse({status: 201, data: initialFetchSize});
      mockAxios.mockResponse({status: 201, data: rowHeight});

      await tick();

      expect(wrapper.update().instance().global.settings[appConstants.SETTING_TIMEZONE]).toEqual(setting);
      expect(wrapper.update().instance().global.settings[appConstants.AUTO_REFRESH_INTERVAL]).toEqual(autoRefreshSetting);
      expect(wrapper.update().instance().global.settings[appConstants.INITIAL_FETCH_SIZE]).toEqual(initialFetchSize);
      expect(wrapper.update().instance().global.settings[appConstants.ROW_HEIGHT]).toEqual(rowHeight);
    });
  });
});
