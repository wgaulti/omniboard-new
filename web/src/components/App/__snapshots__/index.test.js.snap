// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render 1`] = `
<div
  className="App"
>
  <Uncontrolled(Navbar)
    bsStyle="default"
    fluid={true}
    inverse={true}
  >
    <NavbarHeader>
      <NavbarBrand>
        <a
          href="#"
        >
          Omniboard 
          <span
            className="version"
          />
        </a>
      </NavbarBrand>
    </NavbarHeader>
    <NavbarCollapse>
      <Nav
        activeKey="default"
        bsClass="nav"
        justified={false}
        pullLeft={true}
        pullRight={false}
        stacked={false}
      />
      <Nav
        bsClass="nav"
        justified={false}
        pullLeft={false}
        pullRight={true}
        stacked={false}
      >
        <NavDropdown
          eventKey={1}
          id="settings"
          title={
            <Glyphicon
              bsClass="glyphicon"
              glyph="cog"
            />
          }
        >
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey={1.1}
            header={false}
            onClick={[Function]}
            test-attr="reset-cache-button"
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="refresh"
            />
              Reset Cache
          </MenuItem>
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey={1.2}
            header={false}
            onClick={[Function]}
            test-attr="manage-config-columns-button"
          >
            +/- Custom Columns
          </MenuItem>
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey={1.3}
            header={false}
            onClick={[Function]}
            test-attr="settings-button"
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="wrench"
            />
              Settings
          </MenuItem>
        </NavDropdown>
      </Nav>
    </NavbarCollapse>
  </Uncontrolled(Navbar)>
  <div
    className="content"
  >
    <ToastContainer
      autoClose={false}
      bodyClassName={null}
      className={null}
      closeButton={
        <CloseButton
          ariaLabel="close"
        />
      }
      closeOnClick={true}
      draggable={true}
      draggablePercent={80}
      hideProgressBar={false}
      newestOnTop={false}
      pauseOnFocusLoss={true}
      pauseOnHover={true}
      position="top-right"
      progressClassName={null}
      progressStyle={null}
      rtl={false}
      style={null}
      toastClassName={null}
      transition={[Function]}
    />
    <RunsTable
      dbInfo={
        Object {
          "key": "default",
          "name": "test_db",
        }
      }
      handleCustomColumnModalClose={[Function]}
      handleSettingsModalClose={[Function]}
      localStorageKey="default|RunsTable|1"
      showCustomColumnModal={false}
      showSettingsModal={false}
      stateFilterKeys={
        Array [
          "dropdownOptions",
          "columnOrder",
          "columnWidths",
          "sort",
          "metricAndCustomColumns",
          "columnNameMap",
          "statusFilterOptions",
          "filters",
          "autoRefresh",
        ]
      }
    />
  </div>
</div>
`;
